node/runtime/src/lib.rs

```diff
@@ -216,32 +216,36 @@ impl Runtime {
         authority_proposals: &mut Vec<AuthorityStake>,
     ) -> Result<Vec<ReceiptTransaction>, String> {
         let sender_account_id = transaction.body.get_originator();
-        if !is_valid_account_id(&sender_account_id) {
-            return Err("Invalid originator account_id".to_string());
-        }
         let sender: Option<Account> =
             get(state_update, &account_id_to_bytes(COL_ACCOUNT, &sender_account_id));
         match sender {
             Some(mut sender) => {
                 if transaction.body.get_nonce() <= sender.nonce {
+                    // in this case, no need to update sender's nonce
                     return Err(format!(
                         "Transaction nonce {} must be larger than sender nonce {}",
                         transaction.body.get_nonce(),
                         sender.nonce,
                     ));
                 }
+                // update nonce regardless of whether the transaction succeeds
+                sender.nonce = transaction.body.get_nonce();
+                set(state_update, &account_id_to_bytes(COL_ACCOUNT, &sender_account_id), &sender);
+                state_update.commit();
```

Comment:
Actually you need to validate signature before you can update the nonce.

The problem might be that the keys are swapped before the transaction is added to the block, so even though the transaction was accepted to the mempool with valid key, the key might be removed before it's processed here. So you need to check signature again before saving nonce.

Not sure how we can update mempool nonce to update this use-case.
@nearmax 

I guess the short-term solution would be to update the nonce even for invalid signature, to avoid mempool issue, but in the long-term mempool tx nonce cache should monitor accounts somehow.

Reply:
Good point. Actually mempool monitors accounts and validate signatures there. So I should just do this nonce update after validating signature. The real question is that, in the case you described, should we put the transaction in the block?

Reply:
I don't see a huge issue with the current solution. It's rare when you can get you TX accepted to a mempool which later fails for wrong signature. As long as we still do verify signature here, it's secure enough.

Reply:
That's fair.

Reply:
Please add a comment before signature verification to explain the reasoning.